### PR TITLE
(454) Turn on strict TypeScript compilation

### PR DESCRIPTION
**Pull request by Gweaton**
## Context

https://trello.com/c/drGFfal5/454-use-strict-typing-for-typescript

## Changes in this PR

This turns on strict TypeScript compilation to ensure we're getting the most Type safety we can while using TypeScript.

To get this to all compile successfully, we add a bunch of guards, type assertions etc.

## Post-merge checklist

- [ ] Have you written an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e) (if necessary)?

### CODE REVIEW

File: server/data/restClientMetricsMiddleware.test.ts [hunk 1]
```diff
@@ -47,13 +47,13 @@ describe('restClientMetricsMiddleware', () => {
...
-      expect(requestHistogramLabelsSpy).toHaveBeenCalledTimes(1)
-      expect(requestHistogramLabelsSpy).toHaveBeenCalledWith('httpbin.org', 'GET', '/', '200')
-      expect(requestHistogramStartSpy).toHaveBeenCalledTimes(1)
-      nock.cleanAll()
+          expect(code).toBe(200)
+          expect(requestHistogramLabelsSpy).toHaveBeenCalledTimes(1)
+          expect(requestHistogramLabelsSpy).toHaveBeenCalledWith('httpbin.org', 'GET', '/', '200')
+          expect(requestHistogramStartSpy).toHaveBeenCalledTimes(1)
+          nock.cleanAll()
+        })
```
Comment (yndajas) [inline server/data/restClientMetricsMiddleware.test.ts:56]: This one is interesting given the expectations shouldn't run until we've finished (a)waiting. I guess there's a possibility that the chained operations fail somewhere and don't get to setting the code?

File: server/utils/azureAppInsights.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ function initialiseAppInsights(): void {
   }
 }
 
-function buildAppInsightsClient(name = defaultName()): TelemetryClient {
+function buildAppInsightsClient(name = defaultName()): TelemetryClient | null {
```
Comment (yndajas) [inline server/utils/azureAppInsights.ts:27]: These feels similar to `Specify that getToken can return null` - maybe they can go together?

File: server/data/tokenVerification.ts [hunk 3]
```diff
@@ -18,7 +18,16 @@ function getApiClientToken(token: string) {
...
+type RequestWithUser = Request & { user: Express.User }
+
+const assertHasUser: (req: Request) => asserts req is RequestWithUser = req => {
+  if (!('user' in req)) {
+    throw new Error('Request object without user found unexpectedly')
+  }
+}
+
 const tokenVerifier: TokenVerifier = async request => {
+  assertHasUser(request)
```
Comment (yndajas) [inline server/data/tokenVerification.ts:30]: This should be importing `assertHasUser` from `typeUtils`, and not redefine the `RequestWithUser` type

**Review by yndajas — COMMENTED** [review]
Just a few small things to sort out (at least the last one)

File: server/data/tokenVerification.ts (hunk 3 above)
Comment (Gweaton) [reply server/data/tokenVerification.ts]: oops

File: server/data/restClientMetricsMiddleware.test.ts (hunk 1 above)
Comment (Gweaton) [reply server/data/restClientMetricsMiddleware.test.ts]: I've redone this and used the `await`ed value which I think feels more readable here [dc8180f](https://github.com/ministryofjustice/hmpps-accredited-programmes-ui/pull/62/commits/dc8180f30781ed5acc311e3e511be7d03a43be6d)

Comment (yndajas) [reply server/data/restClientMetricsMiddleware.test.ts]: Nice!

File: server/data/tokenVerification.ts (hunk 3 above)
Comment (yndajas) [reply server/data/tokenVerification.ts]: Probably my mistake!

**Review by yndajas — APPROVED** [review]

**Pull request merged**